{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "78bbf917",
   "metadata": {},
   "source": [
    "# File for Playing Around With Thesis Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "49f1612e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m  Activating\u001b[22m\u001b[39m project at `c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1`\n"
     ]
    }
   ],
   "source": [
    "using Pkg\n",
    "\n",
    "# Activate the package repo itself\n",
    "Pkg.activate(\"..\")\n",
    "\n",
    "# Make sure deps are installed\n",
    "Pkg.instantiate()\n",
    "\n",
    "using ImageView, Phylo, Plots, FASTX, MolecularEvolution, CodonMolecularEvolution, JLD2, MCMCChains, StatsPlots, MCMCDiagnosticTools, Measures, BenchmarkTools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4e68e15e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Step 1: Initialization.\n",
      "Step 2: Optimizing global codon model parameters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Warning: XM_015126958_Macaca_mulatta{G2} on tree but not found in names.\n",
      "└ @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\utils\\misc.jl:102\n"
     ]
    },
    {
     "ename": "MethodError",
     "evalue": "MethodError: no method matching identity!(::LazyPartition{NucleotidePartition})\nThe function `identity!` exists, but no method is defined for this combination of argument types.\n\nClosest candidates are:\n  identity!(!Matched::GaussianPartition)\n   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\continuous_models\\gaussian_partition.jl:61\n  identity!(!Matched::CATPartition{PType}) where PType<:DiscretePartition\n   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\compound_models\\cat.jl:36\n  identity!(!Matched::SWMPartition)\n   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\compound_models\\swm.jl:77\n  ...\n",
     "output_type": "error",
     "traceback": [
      "MethodError: no method matching identity!(::LazyPartition{NucleotidePartition})\n",
      "The function `identity!` exists, but no method is defined for this combination of argument types.\n",
      "\n",
      "Closest candidates are:\n",
      "  identity!(!Matched::GaussianPartition)\n",
      "   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\continuous_models\\gaussian_partition.jl:61\n",
      "  identity!(!Matched::CATPartition{PType}) where PType<:DiscretePartition\n",
      "   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\compound_models\\cat.jl:36\n",
      "  identity!(!Matched::SWMPartition)\n",
      "   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\compound_models\\swm.jl:77\n",
      "  ...\n",
      "\n",
      "\n",
      "Stacktrace:\n",
      "  [1] _broadcast_getindex_evalf\n",
      "    @ .\\broadcast.jl:678 [inlined]\n",
      "  [2] _broadcast_getindex\n",
      "    @ .\\broadcast.jl:651 [inlined]\n",
      "  [3] getindex\n",
      "    @ .\\broadcast.jl:610 [inlined]\n",
      "  [4] copy\n",
      "    @ .\\broadcast.jl:911 [inlined]\n",
      "  [5] materialize\n",
      "    @ .\\broadcast.jl:872 [inlined]\n",
      "  [6] uninformative_message!(message::Vector{Partition})\n",
      "    @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\utils\\misc.jl:74\n",
      "  [7] populate_tree!(tree::FelNode, starting_message::Vector{LazyPartition{NucleotidePartition}}, names::Vector{String}, data::Vector{String}; init_all_messages::Bool, tolerate_missing::Int64, leaf_name_transform::CodonMolecularEvolution.var\"#290#294\")\n",
      "    @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\utils\\misc.jl:104\n",
      "  [8] populate_tree!(tree::FelNode, starting_partition::LazyPartition{NucleotidePartition}, names::Vector{String}, data::Vector{String}; init_all_messages::Bool, tolerate_missing::Int64, leaf_name_transform::Function)\n",
      "    @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\utils\\misc.jl:132\n",
      "  [9] populate_tree!\n",
      "    @ C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\utils\\misc.jl:123 [inlined]\n",
      " [10] optimize_nuc_mu(seqnames::Vector{String}, seqs::Vector{String}, tree::FelNode; leaf_name_transform::Function, genetic_code::MolecularEvolution.GeneticCode, optimize_branch_lengths::Bool)\n",
      "    @ CodonMolecularEvolution c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\shared\\shared.jl:264\n",
      " [11] optimize_nuc_mu\n",
      "    @ c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\shared\\shared.jl:254 [inlined]\n",
      " [12] difFUBAR_global_fit_2steps(seqnames::Vector{String}, seqs::Vector{String}, tree::FelNode, leaf_name_transform::Function, code::MolecularEvolution.GeneticCode; verbosity::Int64, optimize_branch_lengths::Bool)\n",
      "    @ CodonMolecularEvolution c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\difFUBAR\\difFUBAR.jl:100\n",
      " [13] difFUBAR_global_fit_2steps\n",
      "    @ c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\difFUBAR\\difFUBAR.jl:96 [inlined]\n",
      " [14] FAMEgrid(seqnames::Vector{String}, seqs::Vector{String}, treestring::String; verbosity::Int64, code::MolecularEvolution.GeneticCode, optimize_branch_lengths::Bool)\n",
      "    @ CodonMolecularEvolution c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\FAME\\FAME.jl:21\n",
      " [15] FAMEgrid(seqnames::Vector{String}, seqs::Vector{String}, treestring::String)\n",
      "    @ CodonMolecularEvolution c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\FAME\\FAME.jl:18\n",
      " [16] top-level scope\n",
      "    @ c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\faffing_about\\jl_notebook_cell_df34fa98e69747e1a8f8a730347b8e2f_W2sZmlsZQ==.jl:5"
     ]
    }
   ],
   "source": [
    "analysis_name = \"output/Ace2\"\n",
    "seqnames,seqs = read_fasta(\"../test/data/Ace2_tiny/Ace2_tiny_tagged.fasta\");\n",
    "treestring = readlines(open(\"../test/data/Ace2_tiny/tiny_tagged_no_bg.tre\"))[1]\n",
    "tags = [\"{G1}\", \"{G2}\"]\n",
    "grid = FAMEgrid(seqnames, seqs, treestring)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "11132e54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Step 1: Initialization.\n",
      "Step 2: Optimizing global codon model parameters.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Warning: XM_015126958_Macaca_mulatta{G2} on tree but not found in names.\n",
      "└ @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\utils\\misc.jl:102\n"
     ]
    },
    {
     "ename": "MethodError",
     "evalue": "MethodError: no method matching identity!(::LazyPartition{NucleotidePartition})\nThe function `identity!` exists, but no method is defined for this combination of argument types.\n\nClosest candidates are:\n  identity!(!Matched::GaussianPartition)\n   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\continuous_models\\gaussian_partition.jl:61\n  identity!(!Matched::CATPartition{PType}) where PType<:DiscretePartition\n   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\compound_models\\cat.jl:36\n  identity!(!Matched::SWMPartition)\n   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\compound_models\\swm.jl:77\n  ...\n",
     "output_type": "error",
     "traceback": [
      "MethodError: no method matching identity!(::LazyPartition{NucleotidePartition})\n",
      "The function `identity!` exists, but no method is defined for this combination of argument types.\n",
      "\n",
      "Closest candidates are:\n",
      "  identity!(!Matched::GaussianPartition)\n",
      "   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\continuous_models\\gaussian_partition.jl:61\n",
      "  identity!(!Matched::CATPartition{PType}) where PType<:DiscretePartition\n",
      "   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\compound_models\\cat.jl:36\n",
      "  identity!(!Matched::SWMPartition)\n",
      "   @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\models\\compound_models\\swm.jl:77\n",
      "  ...\n",
      "\n",
      "\n",
      "Stacktrace:\n",
      "  [1] _broadcast_getindex_evalf\n",
      "    @ .\\broadcast.jl:678 [inlined]\n",
      "  [2] _broadcast_getindex\n",
      "    @ .\\broadcast.jl:651 [inlined]\n",
      "  [3] getindex\n",
      "    @ .\\broadcast.jl:610 [inlined]\n",
      "  [4] copy\n",
      "    @ .\\broadcast.jl:911 [inlined]\n",
      "  [5] materialize\n",
      "    @ .\\broadcast.jl:872 [inlined]\n",
      "  [6] uninformative_message!(message::Vector{Partition})\n",
      "    @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\utils\\misc.jl:74\n",
      "  [7] populate_tree!(tree::FelNode, starting_message::Vector{LazyPartition{NucleotidePartition}}, names::Vector{String}, data::Vector{String}; init_all_messages::Bool, tolerate_missing::Int64, leaf_name_transform::CodonMolecularEvolution.var\"#246#250\")\n",
      "    @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\utils\\misc.jl:104\n",
      "  [8] populate_tree!(tree::FelNode, starting_partition::LazyPartition{NucleotidePartition}, names::Vector{String}, data::Vector{String}; init_all_messages::Bool, tolerate_missing::Int64, leaf_name_transform::Function)\n",
      "    @ MolecularEvolution C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\utils\\misc.jl:132\n",
      "  [9] populate_tree!\n",
      "    @ C:\\Users\\toven\\.julia\\packages\\MolecularEvolution\\rRKnS\\src\\utils\\misc.jl:123 [inlined]\n",
      " [10] optimize_nuc_mu(seqnames::Vector{String}, seqs::Vector{String}, tree::FelNode; leaf_name_transform::Function, genetic_code::MolecularEvolution.GeneticCode, optimize_branch_lengths::Bool)\n",
      "    @ CodonMolecularEvolution c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\shared\\shared.jl:264\n",
      " [11] optimize_nuc_mu\n",
      "    @ c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\shared\\shared.jl:254 [inlined]\n",
      " [12] difFUBAR_global_fit_2steps(seqnames::Vector{String}, seqs::Vector{String}, tree::FelNode, leaf_name_transform::Function, code::MolecularEvolution.GeneticCode; verbosity::Int64, optimize_branch_lengths::Bool)\n",
      "    @ CodonMolecularEvolution c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\difFUBAR\\difFUBAR.jl:100\n",
      " [13] difFUBAR_global_fit_2steps\n",
      "    @ c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\difFUBAR\\difFUBAR.jl:96 [inlined]\n",
      " [14] FLAVORgrid(seqnames::Vector{String}, seqs::Vector{String}, treestring::String; verbosity::Int64, code::MolecularEvolution.GeneticCode, optimize_branch_lengths::Bool)\n",
      "    @ CodonMolecularEvolution c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\FLAVOR\\FLAVOR.jl:22\n",
      " [15] FLAVORgrid(seqnames::Vector{String}, seqs::Vector{String}, treestring::String)\n",
      "    @ CodonMolecularEvolution c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\src\\FLAVOR\\FLAVOR.jl:19\n",
      " [16] top-level scope\n",
      "    @ c:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\faffing_about\\jl_notebook_cell_df34fa98e69747e1a8f8a730347b8e2f_W3sZmlsZQ==.jl:5"
     ]
    }
   ],
   "source": [
    "analysis_name = \"output/Ace2\"\n",
    "seqnames,seqs = read_fasta(\"../test/data/Ace2_tiny/Ace2_tiny_tagged.fasta\");\n",
    "treestring = readlines(open(\"../test/data/Ace2_tiny/tiny_tagged_no_bg.tre\"))[1]\n",
    "tags = [\"{G1}\", \"{G2}\"]\n",
    "flavorgrid = FLAVORgrid(seqnames, seqs, treestring)\n",
    "grid = FAMEgrid(seqnames, seqs, treestring)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2203867f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36m\u001b[1mProject\u001b[22m\u001b[39m CodonMolecularEvolution v0.1.0\n",
      "\u001b[32m\u001b[1mStatus\u001b[22m\u001b[39m `C:\\Users\\toven\\Documents\\tove-code\\MurrellLab\\CodonMolecularEvolution.jl-1\\Project.toml`\n",
      "  \u001b[90m[9f975960] \u001b[39mMolecularEvolution v0.2.7\n"
     ]
    }
   ],
   "source": [
    "\n",
    "Pkg.status(\"MolecularEvolution\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4953d198",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.11.5",
   "language": "julia",
   "name": "julia-1.11"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
